components/modules/otahttp.c

```diff
@@ -0,0 +1,205 @@
+{
+    int i = 0;
+
+    while (http_response[i] != 0 && i + 3 < total_len) {
+      if(http_response[i++] == '\r' && http_response[i++] == '\n' && http_response[i++] == '\r' && http_response[i++] == '\n') {
+        return i;
+      }
+    }
+
+    return -1;
+}
+
+static int otahttp_download( lua_State *L )
+{
+  if(!ota_started) {
+    OTAHTTP_ERROR(L, "Error: OTA not started");
+  }
+
+  int image_length = 0;
+  char http_response[BUFFSIZE + 1];
+  char write_data[BUFFSIZE + 1];
+
+  esp_err_t err;
+  int stack = 0;
+
+  size_t domain_length = 0;
+  const char* domain = luaL_checklstring(L, ++stack, &domain_length);
+  uint16_t port = luaL_checkinteger(L, ++stack);
+  size_t path_length = 0;
+  const char* path = luaL_checklstring(L, ++stack, &path_length);
+
+  /*connect to http server*/
+  int socket_id = connect_to_http_server(domain, port);
+  if (socket_id < 0) {
+    OTAHTTP_ERROR(L, "Connect to http server failed! errno=%d", socket_id);
+  } else {
+    ESP_LOGI(TAG, "Connected to http server");
+  }
+
+  /*send GET request to http server*/
+  char http_request[64] = {0};
+  sprintf(http_request, "GET %s HTTP/1.1\r\nHost: %s:%d \r\n\r\n", path, domain, port);
+
+  int res = send(socket_id, http_request, strlen(http_request), 0);
+  if (res == -1) {
+    OTAHTTP_ERROR(L, "Send GET request to server failed");
+  } else {
+    ESP_LOGI(TAG, "Send GET request to server succeeded");
+  }
+
+  bool found_body = false, flag = true;
+  int packet_counter = 0;
+
+  /* Deal with all receive packets */
+  while (flag) {
+    memset(&http_response, 0, TEXT_BUFFSIZE);
+    memset(&write_data, 0, BUFFSIZE);
+
+    int packet_length = recv(socket_id, &http_response, TEXT_BUFFSIZE, 0);
+    if (packet_length < 0) { /* Receive error */
+      OTAHTTP_ERROR(L, "Error: receive data error! errno=%d", packet_length);
+    } else if(packet_length == 0) { /* No more data, we´re done */
+      flag = false;
+      ESP_LOGI(TAG, "Connection closed, all packets received");
+      close(socket_id);
+      break;
+    }
+
+    packet_counter++;
+
+    int body_offset = 0;
+    if(!found_body) {
+      body_offset = find_body_offset(http_response, packet_length);
+      if(body_offset == -1)
+        continue;
+
+      found_body = true;
+    }
+    int length = packet_length - body_offset;
+
+    memcpy(&write_data, &(http_response[body_offset]), length);
+    err = esp_ota_write( ota_update_handle, (const void *)&write_data, length);
+    if (err != ESP_OK) {
+      close(socket_id);
+      OTAHTTP_ERROR(L, "Error: esp_ota_write failed! err=0x%x", err);
+    }
+
+    image_length += length;
+    if(packet_counter % 100 == 0) {
+      ESP_LOGI(TAG, "Have written image length %d", image_length);
+    }
+  }
+
+  if(!found_body) {
+    OTAHTTP_ERROR(L, "Error: did not found body!");
+  } else {
+    ESP_LOGI(TAG, "Total Write binary data length : %d", image_length);
+
+    lua_pushboolean (L, true);
```

Comment:
...and rather than returning the success flag here, it'd be done by invoking a Lua callback. A callback of the canonical format `callback(err, other_data)` would be ideal, where `err` is null if there was no error.